File: 1주차/sehyeon/광물캐기.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(vector<int> picks, vector<string> minerals)
{
  int answer = 0;
  int cnt = 0;
  for (auto i : picks)
    cnt += i;
  if (minerals.size() > cnt * 5)
    minerals.erase(minerals.begin() + (cnt * 5), minerals.end());

  priority_queue<char> min;
  priority_queue<string> sort_minerals;

  for (int i = 0; i < minerals.size(); i++)
  {
    if (minerals[i] == "diamond")
      min.push('a');
    else if (minerals[i] == "iron")
      min.push('b');
    else
      min.push('c');
    if (i % 5 == 4 || i == minerals.size() - 1)
    {
      while (min.size() < 5)
        min.push('d');
      string str = "";
      while (!min.empty())
      {
        str += min.top();
        min.pop();
      }
      sort_minerals.push(str);
    }
  }

  for (int i = 0; i < picks.size(); i++)
  {
    while (picks[i]-- && !sort_minerals.empty())
    {
      string tmp = sort_minerals.top();
      if (i == 0)
      {
        for (int j = 0; j < tmp.length(); j++)
          if (tmp[j] != 'd')
            answer++;
      }
      else if (i == 1)
      {
        for (int j = 0; j < tmp.length(); j++)
        {
          if (tmp[j] == 'a')
            answer += 5;
          else if (tmp[j] != 'd')
            answer++;
        }
      }
      else
      {
        for (int j = 0; j < tmp.length(); j++)
        {
          if (tmp[j] == 'a')
            answer += 25;
          else if (tmp[j] == 'b')
            answer += 5;
          else if (tmp[j] == 'c')
            answer++;
        }
      }
      sort_minerals.pop();
    }
  }

  return answer;
}

File: 1주차/sehyeon/광물캐기.js
function solution(picks, minerals) {
  var answer = 0;
  let cnt = 0;
  let sort_str = [];
  picks.forEach((ele) => cnt += ele);
  minerals = minerals.slice(0, cnt * 5);

  let str = [];
  minerals.forEach((ele, index) => {
    if (ele === "diamond")
      str.push('a');
    else if (ele === "iron")
      str.push('b');
    else
      str.push('c');
    if (index % 5 === 4 || index == minerals.length - 1) {
      if (str.length !== 5) {
        while (str.length < 5)
          str.push('d');
      }
      str.sort();
      let s = '';
      str.forEach((ele) => {
        s += ele;
      })
      sort_str.push(s);
      str = [];
    }
  })
  sort_str.sort();

  console.log(sort_str);

  picks.forEach((pick, idx) => {
    while (pick-- && sort_str.length) {
      if (idx == 0) {
        let tmp = sort_str[0];
        for (let i = 0; i < tmp.length; i++) {
          if (tmp[i] !== 'd')
            answer++;
        }
        sort_str.shift();
      } else if (idx == 1) {
        let tmp = sort_str[0];
        for (let i = 0; i < tmp.length; i++) {
          if (tmp[i] === 'a') {
            answer += 5;
          } else if (tmp[i] != 'd') {
            answer++;
          }
        }
        sort_str.shift();
      } else {
        let tmp = sort_str[0];
        for (let i = 0; i < tmp.length; i++) {
          if (tmp[i] === 'a') {
            answer += 25;
          } else if (tmp[i] === 'b') {
            answer += 5;
          } else if (tmp[i] == 'c') {
            answer++;
          }
        }
        sort_str.shift();
      }
    }
  })
  return answer;
}